pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+
+    f1: (int, str) -> bool
+    f1: Callable[[int, str], bool]
+
+    f2: (...) -> bool
+    f2: Callable[..., bool]
+
+    f3: async (str) -> str
+    f3: Callable[[str], Awaitable[str]]
+
+    f4: (**P) -> bool
+    f4: Callable[P, bool]
+
+    f5: (int, **P) -> bool
+    f5: Callable[Concatenate[int, P], bool]
+
+    f6: (*Ts) -> bool
+    f6: Callable[[*Ts], bool]
+
+    f7: (int, *Ts, str) -> bool
+    f7: Callable[[int, *Ts, str], bool]
+
+
+Grammar and AST
+---------------
+
+The proposed new syntax can be described by these AST changes ::
+
+    expr = <prexisting_expr_kinds>
+         | AsyncCallableType(callable_type_arguments args, expr returns)
+         | CallableType(callable_type_arguments args, expr returns)
+
+    callable_type_arguments = AnyArguments
+                            | ArgumentsList(expr* posonlyargs)
+                            | Concatenation(expr* posonlyargs, expr param_spec)
+
+
+Here are our proposed changes to the `Python Grammar
+<https://docs.python.org/3/reference/grammar.htm>`::
+
+    expression:
+        | disjunction disjunction 'else' expression
+        | callable_type_expression
+        | disjunction
+        | lambdef
+
+    callable_type_expression:
+        | callable_type_arguments '->' expression
+        | ASYNC callable_type_arguments '->' expression
+
+    callable_type_arguments:
+        | '(' '...' [','] ')'
+        | '(' callable_type_positional_argument*  ')'
+        | '(' callable_type_positional_argument* callable_type_param_spec ')'
+
+    callable_type_positional_argument:
+        | !'...' expression ','
+        | !'...' expression &')'
+
+    callable_type_param_spec:
+        | '**' expression ','
+        | '**' expression &')'
+
+
+
+If PEP 646 is accepted, we intend to include support for unpacked
+types by modifying the grammar for
+``callable_type_positional_argument`` as follows::
+
+    callable_type_positional_argument:
+        | expression ','
+        | expression &')'
+        | '*' expression ','
+        | '*' expression &')'
+
+
+Implications of the Grammar
+---------------------------
+
+
+Precedence of ->
+~~~~~~~~~~~~~~~~
+
+
+``->`` binds less tightly than other operators, both inside types and
+in function signatures, so the following two callable types are
+equivalent::
+
+    (int) -> str | bool
+    (int) -> (str | bool)
+
+
+``->`` associates to the right, both inside types and in function
+signatures. So the following pairs are equivalent::
+
+    (int) -> (str) -> bool
+    (int) -> ((str) -> bool)
+
+    def f() -> (int, str) -> bool: pass
+    def f() -> ((int, str) -> bool): pass
+
+    def f() -> (int) -> (str) -> bool: pass
+    def f() -> ((int) -> ((str) -> bool)): pass
+
+
+Because operators bind more tightly than ``->``, parentheses are
+required whenever an arrow type is intended to be inside an argument
+to an operator like ``|``::
+
+    (int) -> bool | () -> bool    # syntax error!
+    (int) -> bool | (() -> bool)  # okay
```

Comment:
So that's a function from (int) to a union type, the right half of which is another function.

Maybe a more useful example would be the union of two function types?